cmd/hepa/consumer.go

```diff
@@ -0,0 +1,130 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	comatproto "github.com/bluesky-social/indigo/api/atproto"
+	"github.com/bluesky-social/indigo/atproto/syntax"
+	"github.com/bluesky-social/indigo/events/schedulers/autoscaling"
+	lexutil "github.com/bluesky-social/indigo/lex/util"
+
+	"github.com/bluesky-social/indigo/events"
+	"github.com/bluesky-social/indigo/repo"
+	"github.com/bluesky-social/indigo/repomgr"
+	"github.com/carlmjohnson/versioninfo"
+	"github.com/gorilla/websocket"
+)
+
+func (s *Server) RunConsumer(ctx context.Context) error {
+
+	// TODO: persist cursor in a database or local disk
+	cur, err := s.ReadLastCursor(ctx)
+	if err != nil {
+		return err
+	}
+
+	dialer := websocket.DefaultDialer
+	u, err := url.Parse(s.bgshost)
+	if err != nil {
+		return fmt.Errorf("invalid bgshost URI: %w", err)
+	}
+	u.Path = "xrpc/com.atproto.sync.subscribeRepos"
+	if cur != 0 {
+		u.RawQuery = fmt.Sprintf("cursor=%d", cur)
+	}
+	s.logger.Info("subscribing to repo event stream", "upstream", s.bgshost, "cursor", cur)
+	con, _, err := dialer.Dial(u.String(), http.Header{
+		"User-Agent": []string{fmt.Sprintf("hepa/%s", versioninfo.Short())},
+	})
+	if err != nil {
+		return fmt.Errorf("subscribing to firehose failed (dialing): %w", err)
+	}
+
+	rsc := &events.RepoStreamCallbacks{
+		RepoCommit: func(evt *comatproto.SyncSubscribeRepos_Commit) error {
+			s.lastSeq = evt.Seq
+			return s.HandleRepoCommit(ctx, evt)
+		},
+		RepoHandle: func(evt *comatproto.SyncSubscribeRepos_Handle) error {
+			s.lastSeq = evt.Seq
+			did, err := syntax.ParseDID(evt.Did)
+			if err != nil {
+				s.logger.Error("bad DID in RepoHandle event", "did", evt.Did, "handle", evt.Handle, "seq", evt.Seq, "err", err)
+				return nil
+			}
+			if err := s.engine.ProcessIdentityEvent(ctx, "handle", did); err != nil {
+				s.logger.Error("processing handle update failed", "did", evt.Did, "handle", evt.Handle, "seq", evt.Seq, "err", err)
+			}
+			return nil
+		},
+		// TODO: other event callbacks as needed
+	}
+
+	// start at higher parallelism (somewhat arbitrary)
+	scaleSettings := autoscaling.DefaultAutoscaleSettings()
+	scaleSettings.Concurrency = 6
+	return events.HandleRepoStream(
+		ctx, con, autoscaling.NewScheduler(
+			scaleSettings,
+			s.bgshost,
+			rsc.EventHandler,
+		),
+	)
+}
+
+// NOTE: for now, this function basically never errors, just logs and returns nil. Should think through error processing better.
+func (s *Server) HandleRepoCommit(ctx context.Context, evt *comatproto.SyncSubscribeRepos_Commit) error {
+
+	logger := s.logger.With("event", "commit", "did", evt.Repo, "rev", evt.Rev, "seq", evt.Seq)
+	logger.Debug("received commit event")
+
+	if evt.TooBig {
+		logger.Warn("skipping tooBig events for now")
+		return nil
+	}
+
+	did, err := syntax.ParseDID(evt.Repo)
```

Comment:
not now, but we should probably make lexgen use the syntax.DID type for fields that have format:did

Reply:
yup! I can't remember if I have tested yet, but the syntax.DID type implements the text marshaling interface, so this should Just Work